[Bug] Nondeterministic Ordering in Generated Docs Topics

### Search before reporting

- [x] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.


### Read release policy

- [x] I understand that [unsupported versions](https://pulsar.apache.org/contribute/release-policy/#supported-versions) don't get bug fixes. I will attempt to reproduce the issue on a supported version of Pulsar client and Pulsar broker.


### User environment

Broker version: 4.2.0-SNAPSHOT
Broker Operating system and hardware type: Linux x86_64 GNU/Linux
Broker Java version: 17.0.16

### Issue Description

The below test compares the generated documentation of `CmdGenerateDocsTest#testGenerateDocs`.  The test assumes a deterministic order of the generated documentation but the order of the listed `help` and `name` options was not guaranteed. Due to two sources of nondeterminism hidden in the test and production code, the options could be listed in a different order. Picocli's `CommandLine` class uses [Class#getDeclaredFields](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getDeclaredFields) internally (via reflection), which has a non-deterministic specification to retrieve the `Option` annotations and orders these options accordingly. Even though picocli supports an order attribute on `Option` annotations, [CommandLine](https://github.com/remkop/picocli/blob/main/src/main/java/picocli/CommandLine.java) never sorts by this order - it just maintains whatever order the reflection API gave it. `CmdGenerateDocs#generateDocument` then iterates over the order that the `args` method returns, with the assumption that the order is deterministic. Unfortunately, the ordering can change due to different environments producing the contents in different orders despite the logical contents being the same. Harmless re-ordering could flip the tests from pass to fail despite the data being semantically the same.
- `org.apache.pulsar.docs.tools.CmdGenerateDocsTest.testGenerateDocs`

I discovered the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool. NonDex systematically detects incorrect tests that rely on non-deterministic behaviors in Java APIs—like assuming order of name/value pairs in json files — by exploring all specification-allowed outcomes. It does this by instrumenting undetermined APIs and randomizing the returned order/behavior within what the spec allows. Failures it exposes reliably indicate flawed assumptions in order that were never promised and should be fixed.

### Error messages

```text
[INFO] Running org.apache.pulsar.docs.tools.CmdGenerateDocsTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 3.273 s <<< FAILURE! -- in org.apache.pulsar.docs.tools.CmdGenerateDocsTest
[ERROR] org.apache.pulsar.docs.tools.CmdGenerateDocsTest.testGenerateDocs -- Time elapsed: 0.007 s <<< FAILURE!
java.lang.AssertionError: 
expected [# test




$ pulsar test options


|Flag|Description|Default|
|---|---|---|
| `-h, --help` | Show this help message|false|
| `-n, --name` | Name|null|

] but found [# test




$ pulsar test options


|Flag|Description|Default|
|---|---|---|
| `-n, --name` | Name|null|
| `-h, --help` | Show this help message|false|

]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.failNotEquals(Assert.java:1577)
	at org.testng.Assert.assertEqualsImpl(Assert.java:149)
	at org.testng.Assert.assertEquals(Assert.java:131)
	at org.testng.Assert.assertEquals(Assert.java:655)
	at org.testng.Assert.assertEquals(Assert.java:665)
	at org.apache.pulsar.docs.tools.CmdGenerateDocsTest.testGenerateDocs(CmdGenerateDocsTest.java:79)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR] org.apache.pulsar.docs.tools.CmdGenerateDocsTest.testGenerateDocs
[INFO]   Run 1: PASS
[ERROR]   Run 2: CmdGenerateDocsTest.testGenerateDocs:79 expected [# test




$ pulsar test options


|Flag|Description|Default|
|---|---|---|
| `-h, --help` | Show this help message|false|
| `-n, --name` | Name|null|

] but found [# test




$ pulsar test options


|Flag|Description|Default|
|---|---|---|
| `-n, --name` | Name|null|
| `-h, --help` | Show this help message|false|

]
[INFO] 
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

### Reproducing the issue

You can replicate the problem with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool command:
```
mvn -pl pulsar-docs-tools -Dtest=org.apache.pulsar.docs.tools.CmdGenerateDocsTest#testGenerateDocs -DnondexRuns=10 -DreuseForks=false edu.illinois:nondex-maven-plugin:2.2.1:nondex -Dcheckstyle.skip=true -DnondexRunsWithoutShuffling=0 
```

### Additional information

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/pulsar-docs-tools/src/main/java/org/apache/pulsar/docs/tools/CmdGenerateDocs.java
+++ b/pulsar-docs-tools/src/main/java/org/apache/pulsar/docs/tools/CmdGenerateDocs.java
@@ -108,7 +108,7 @@ public class CmdGenerateDocs implements Callable<Integer> {
                     sb.append("```shell\n$ ");
                     sb.append(this.commander.getCommandName()).append(" ");
                     sb.append(module).append(" ").append(subK).append(" options").append("\n```\n\n");
-                    List<ArgSpec> argSpecs = subV.getCommandSpec().args();
+                    List<ArgSpec> argSpecs = getSortedArgs(subV.getCommandSpec().args());
                     if (argSpecs.size() > 0) {
                         sb.append("|Flag|Description|Default|\n");
                         sb.append("|---|---|---|\n");
@@ -128,7 +128,7 @@ public class CmdGenerateDocs implements Callable<Integer> {
             sb.append(" options").append("\n```").append("\n\n");
             sb.append("|Flag|Description|Default|\n");
             sb.append("|---|---|---|\n");
-            List<ArgSpec> argSpecs = commander.getCommandSpec().args();
+            List<ArgSpec> argSpecs = getSortedArgs(commander.getCommandSpec().args());
             argSpecs.forEach(option -> {
                 if (option.hidden() || !(option instanceof OptionSpec)) {
                     return;
@@ -142,6 +142,27 @@ public class CmdGenerateDocs implements Callable<Integer> {
             });
         }
         return sb.toString();
+    }
+
+    /**
+     * Returns a sorted copy of the argument specifications list, ordered by the
+     * OptionSpec order attribute for option specs. Non-option arguments
+     * retain their relative ordering.
+     *
+     * @param args the list of argument specifications to sort
+     * @return a new sorted list
+     */
+    private List<ArgSpec> getSortedArgs(List<ArgSpec> args) {
+        List<ArgSpec> sorted = new ArrayList<>(args);
+        sorted.sort((a, b) -> {
+            if (a instanceof OptionSpec && b instanceof OptionSpec) {
+                OptionSpec optA = (OptionSpec) a;
+                OptionSpec optB = (OptionSpec) b;
+                return Integer.compare(optA.order(), optB.order());
+            }
+            return 0;
+        });
+        return sorted;
     }
 
     @Override

--- a/pulsar-docs-tools/src/test/java/org/apache/pulsar/docs/tools/CmdGenerateDocsTest.java
+++ b/pulsar-docs-tools/src/test/java/org/apache/pulsar/docs/tools/CmdGenerateDocsTest.java
@@ -29,10 +29,10 @@ public class CmdGenerateDocsTest {
 
     @Command
     public class Arguments {
-        @Option(names = {"-h", "--help"}, description = "Show this help message")
+        @Option(names = {"-h", "--help"}, description = "Show this help message", order = 0)
         private boolean help = false;
 
-        @Option(names = {"-n", "--name"}, description = "Name")
+        @Option(names = {"-n", "--name"}, description = "Name", order = 1)
         private String name;
     }
 

